### ISSUE TITLE
Running exificient on Android

### ISSUE THREAD

**Issue by bangarharshit** [EXIficient/exificient#1]
Hi Guys, I am trying to see how I can make it run on android? Any pointers about what are the classes I can ignore. Currently I am using standard xml pull parser and I am trying to replace xml with exi?

**Comment by danielpeintner** [comment]
Hi,

Sorry for delay in our answer... Christmas time..

Back to your question. A while ago (at the time Android 2.3 was current) I ported the EXIficient code to Android. At that time the only issue I encountered was problems with the Xerces-J library that is required to generate schema-informed EXI grammars from XML schema documents.

That said, I would recommend removing those dependencies and create  com.siemens.ct.exi.grammars.grammar.SchemaInformedGrammar's differently if needed.

A list of those classes are (maybe some more)
com.siemens.ct.exi.grammars.XSDGrammarsBuilder
com.siemens.ct.exi.grammars.EXIContentModelBuilder
org.apache.xerces.impl.xpath.regex.EXIRegularExpression
and the Xerces-J library itself.

We plan to separate the XSD part from the core library. You may want to wait for that so that you can start from EXIficient without any external dependencies which might be even easier.

Hope this helps,

-- Daniel

P.S. Please keep us informed
P.P.S. Please let us know if we change our code to support your efforts

**Comment by bangarharshit** [comment]
Hey, Sorry for the delay in answer since I was on holiday.

I am trying to implement exi for xmpp - http://xmpp.org/extensions/xep-0322.html. I am implementing it for this project - https://github.com/bangarharshit/Smack. EXI is certainly the most exciting thing in XMPP in recent years but there are no servers to talk to. So currently I am making fixes for one of the servers openfire to talk to exi. Once that will be done then I will be making fixes for client.

In any case it will be great if it is possible to use exi on android since soon there will be many android clients which will be implementing it.

**Comment by danielpeintner** [comment]
Hi,

as promised we separated the various EXIficient parts. We now have

- https://github.com/EXIficient/exificient-core for the core parts
- https://github.com/EXIficient/exificient-grammars for creating EXI grammars from XML schema and
- https://github.com/EXIficient/exificient as wrapper around the previous two projects with SAX, StAX, DOM API support

In the Android case I think it should be possible to use the core part and maybe also pick one API that fits best.

Please let me know if it works or whether you have comments/suggestions.

**Comment by LukashenkoEvgeniy** [comment]
Please let me know if it works on Android?

**Comment by danielpeintner** [comment]
Hi,

because I am curious I tried to create an Android project on my own with exificient-core.
I can encode and decode EXI documents just fine.

Note: I did not test DOM nor the support of XML schemas. DOM support seems easy but I doubt that the XML Schema processors Xerces2-J is supported. This means if schema support is needed the grammars needs to be generated elsewhere... 

FYI: It runs at least with android:minSdkVersion 8.

Hope this helps,

-- Daniel

**Comment by Flowdalic** [comment]
> Note: I did not test DOM nor the support of XML schemas. DOM support seems easy but I doubt that the XML Schema processors Xerces2-J is supported. This means if schema support is needed the grammars needs to be generated elsewhere...

If I get that right, that means what we can currently  use exificient on Android only in schema uninformed mode. Is that correct?

I also see that only Stax, DOM and SAX are supported. While [I do have plans to move from XPP3 to Stax in Smack (SMACK-591)](https://issues.igniterealtime.org/browse/SMACK-591), it would be great if the XPP3 interface would also be supported. Simply because XPP3 is a native part of Android and thus no third party dependencies would be required to get exificient running on Android. Considering the target devices running Android it is also a plus if we can avoid a (sometimes heavy-weight) dependency and resort to native Android API. As consequence, grammar generation using XPP3 would also be desirable.

**tl;dr:** It would be great for EXI on Android if exificent would (optionally) use XPP3 for grammar generation and expose an API class implementing the XPP3 interface. This would also make the integration of exificent in Smack, for XMPP using EXI on Android, very smooth.

**Comment by danielpeintner** [comment]
exificient-core (which can be used on Android) supports schema-informed streams (by using schema-informed EXI grammars). 

Having said that, there is currently no way to parse an XML schema on android and creating schema-informed EXI grammars out of it (I have heard of plans porting Xerces-J to Android but I don't think this made it, correct?).

Hence, one can serialize the EXI grammar classes in one or the other way and parse it again. Currently there is no standardized way in EXI of doing that.  For the JavaScript implementation (see https://github.com/EXIficient/exificient-grammars) I have written a Grammars2JSON.java serializer that writes out JSON which than can be interpreted by the JavaScript implementation.

A similar approach could be used for Android. 
Other ideas are also welcome.
Serializable could be another approach (not sure if Android supports that)

Comments/ideas welcome!

**Comment by Flowdalic** [comment]
I think the most important thing is that the schema-informed EXI grammar generation is done by using only the Android API, so that no dependencies, like Xerces, are required.

XMPP clients receive the schema as XML according to XEP-0322. Therefore, with my hat as XMPP client library developer on, I want an EXI API which is able to use that XML in hand for schema-informed streams, without third party dependencies.

I tried to look for the code which generates the grammars from XML (using Xerces as far as I understand). But unfortunately without any success. Where does that magic happen?

> Serializable could be another approach (not sure if Android supports that)

Serializable as in `java.io.Serializable`? Yes, that is supported on Android. Not sure how that would help though.

**Comment by danielpeintner** [comment]
> I think the most important thing is that the schema-informed EXI grammar generation is done by using only the Android API, so that no dependencies, like Xerces, are required.

Makes sense.

> XMPP clients receive the schema as XML according to XEP-0322. Therefore, with my hat as XMPP client library developer on, I want an EXI API which is able to use that XML in hand for schema-informed streams, without third party dependencies.

Right. I think there is just a set of different schemas. Or put differently there is just a set of schemas an Android client needs to understand. Is this correct?

IF so... one could pre-generate the EXI grammars which are needed in

1.  a form the Android clients can read (easy consumable intermediary form) or
2.  Serialize the EXI grammars (on another machine) via  java.io.Serializable so that the Android client can read it.

> I tried to look for the code which generates the grammars from XML (using Xerces as far as I understand). But unfortunately without any success. Where does that magic happen?

The code for creating EXI grammars is done in https://github.com/EXIficient/exificient-grammars. The project is not very huge but it uses Xerces-J. I would not recommend doing it from scratch without any help of an XML schema parser library. Hence I propose the two approaches mentioned before.

Does that make sense?
Do you have other proposals in mind?

**Comment by Flowdalic** [comment]
> Right. I think there is just a set of different schemas. Or put differently there is just a set of schemas an Android client needs to understand. Is this correct?

Actually, I was wrong. After reading XEP-0322 again, it appears that clients *never* retrieve the schema from the server (I was confused by the "Downloading new schema files on server" section). So there is, considering the current state of the XEP, no need for clients to dynamically generate EXI grammars from XML schemas. (Which eventually makes sense, because clients know a priori which XMPP extensions they understand, and thus which EXI grammars are required).

But yes, this means that pre-generated EXI grammars are sufficient for XMPP clients.

And since exificient-core declares no runtime dependency, the only item left on my wishlist is XPP3 support. :)

**Comment by danielpeintner** [comment]
Hi @Flowdalic ,

I did create a dedicated issue for XPP3 support (see https://github.com/EXIficient/exificient/issues/4).
I can't promise it will happen soon though.

An alternative is to switch with Smack to XmlStreamReader as you suggested (https://issues.igniterealtime.org/browse/SMACK-591) OR to create an XPP3 wrapper on your own. You being familiar with XPP3 and looking at our existing SAX or StAX API should be able to do that pretty fast...

**Comment by danielpeintner** [comment]
XPP3 support has been implemented and discussed in https://github.com/EXIficient/exificient/issues/4
